ocra_tool/ocra_tool.8
.\" Copyright (c) 2014, 2018 Stefan Grundmann
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"
.Dd March 25, 2018
.Dt OCRA_TOOL 8
.Os
.Sh NAME
.Nm ocra_tool
.Nd create and view OCRA credential files
.Sh SYNOPSIS
.Nm
.Cm info
.Fl f
.Ar credential_file
.Nm
.Cm init
.Fl f
.Ar credential_file
.Fl s
.Ar suite_string
.Fl k
.Ar key
.Op Fl c Ar counter
.Op Fl p Ar pin | Fl P Ar pin_hash
.Op Fl q Ar kill_pin | Fl Q Ar kill_pin_hash
.Op Fl w Ar counter_window
.Op Fl t Ar timestamp_offset
.Nm
.Cm sync
.Fl f
.Ar credential_file
.Fl c
.Ar challenge
.Fl r
.Ar response
.Fl v
.Ar second_response
.Sh DESCRIPTION
The
.Nm
utility is used to initialize and view OCRA credential files used by the
.Xr pam_ocra 8 authentication service module for PAM.
.Bl -tag -width ".Cm info"
.It Cm info
Show content of OCRA credential file.
.Bl -tag -width ".FL f"
.It Fl f Ar credential_file
the OCRA credential file.
.El
.It Cm init
Initialize OCRA credential file.
.Bl -tag -width ".FL f"
.It Fl f Ar credential_file
the OCRA credential file.
.It Fl s Ar suite_string
OCRA suite string.
.It Fl k Ar key
specified as hexadecimal string.
.It Fl c Ar counter
If the
.Ar suite_string
requires a counter parameter,
.Ar counter
is the initial counter value. If
.Ar counter
begins with '0x' it will be interpreted as hexadecimal number.
.It Fl p Ar pin | Fl P Ar pin_hash
If the
.Ar suite_string
requires a pin-hash parameter, it is either set as hexadecimal string
.Ar pin_hash
or calculated from
.Ar pin
using the pin-hash algorithm in
.Ar suite_string .
.It Fl q Ar kill_pin | Fl Q Ar kill_pin_hash
If the
.Ar suite_string
requires a pin-hash parameter, it is optional to provide a kill pin. If this
pin is used, the the authentication fails, all future authentications also fail.
The kill pin is set as hexadecimal string
.Ar kill_pin_hash
or calculated from
.Ar kill_pin
using the pin-hash algorithm in
.Ar suite_string .
.It Fl w Ar counter_window
If the
.Ar suite_string
requires a counter parameter,
.Ar counter_window
specifies the maximum number of verify attempts
.Xr pam_ocra 8
will make (while incrementing the counter value).
This parameter is optional.
.It Fl t Ar timestamp_offset
If the
.Ar suite_string
requires a timestamp parameter,
.Ar timestamp_offset
specifies the number of timestamp steps
.Xr pam_ocra 8
will make while verifying a response. The verify process will start at
(current_timestamp -
.Ar timestamp_offset )
and end at (current_timestamp +
.Ar timestamp_offset ) .
This parameter is optional.
.El
.It Cm sync
Synchronize the counter with two responses from the OTP device. Based on a
user selectable challenge and two subsequent responses it is possible to brute
force the counter in the OTP device.
When the counter is found, the database is updated. The counter is searched from
0 to UINT_MAX.
.Bl -tag -width ".FL f"
.It Fl f Ar credential_file
the OCRA credential file.
.It Fl c Ar challenge
Challenge entered in the OTP device.
.It Fl r Ar response
First response for the given challenge.
.It Fl v Ar second response
Second response for the given challenge.
.El
.El
.Sh EXAMPLE
$ ocra_tool init -f ~/.ocra -s OCRA-1:HOTP-SHA1-6:C-QN08-PSHA1 \\
.sp 0
        -k 00112233445566778899aabbccddeeff00112233 \\
.sp 0
        -c 10 -p 1234 -w 50
.sp 1
$ ocra_tool info -f ~/.ocra
.sp 0
suite:          OCRA-1:HOTP-SHA1-6:C-QN08-PSHA1
.sp 0
key:            0x00112233445566778899aabbccddeeff00112233
.sp 0
counter:        0x000000000000000a
.sp 0
counter_window: 50
.sp 0
pin_hash:       0x7110eda4d09e062aa5e4a390b0a572ac0d2c0220
.sp 0
kill_pin_hash:  NOT SET
.sp 0
kill pin used:  false
.Lp
.Sh FILES
.Bl -tag -width Ds -compact
.It Pa ~/.ocra
OCRA credential file
.El
.Sh SEE ALSO
.Xr pam_ocra 8
.Sh AUTHOR
.Nm
and this manual page were developed by Stefan Grundmann.


ocra_tool/ocra_tool.c
/*
 * Copyright (c) 2014,2018 Stefan Grundmann
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <stdint.h>
#include <stdio.h>
#include <unistd.h>
#include <inttypes.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <err.h>
#include <errno.h>
#include <sysexits.h>

#include <db.h>
#include <fcntl.h>

#include <openssl/evp.h>
#include <openssl/err.h>

#include "rfc6287.h"

#define KEY(k, s) memcpy(k.data = K_buf, s, k.size = sizeof(s));
#define VALUE(v, s, z) memcpy(v.data = V_buf, s, v.size = z);

static char K_buf[32];
static char V_buf[254];

static void
pin_hash(const ocra_suite * ocra, const char *pin, uint8_t **P, size_t *P_l)
{
	unsigned int s;
	EVP_MD_CTX ctx;

	*P = NULL;
	*P_l = mdlen(ocra->P_alg);
	EVP_MD_CTX_init(&ctx);

	if (NULL == (*P = (uint8_t *)malloc(*P_l)))
		err(EX_OSERR, "malloc() failed");

	if ((1 != EVP_DigestInit(&ctx, evp_md(ocra->P_alg))) ||
	    (1 != EVP_DigestUpdate(&ctx, pin, strlen(pin))) ||
	    (1 != EVP_DigestFinal(&ctx, *P, &s)) ||
	    (s != *P_l))
		errx(EX_OSERR, "pin_hash() failed: %s",
		    ERR_error_string(ERR_get_error(), NULL));
	EVP_MD_CTX_cleanup(&ctx);
}

static int
parse_counter(const char *in, uint64_t *C)
{
	char *stopped;
	int base = strncmp("0x", in, 2) ? 10 : 16;

	if ('-' == in[0])
		return -1;
	*C = strtouq(in, &stopped, base);
	if (ULLONG_MAX == *C || 0 == *C) {
		if (errno)
			return -1;
	}
	if (*stopped)
		return -1;
	return 0;
}

static int
parse_num(const char *in)
{
	char *stopped;
	int x = (int)strtol(in, &stopped, 10);

	if (*stopped || (0 > x))
		return -1;
	return x;
}

static int
from_hex(const char *in, uint8_t **out, size_t len)
{
	uint32_t i;

	if (0 == strncmp("0x", in, 2))
		in += 2;
	if (strlen(in) != (len * 2))
		return -1;
	if (NULL == (*out = (uint8_t *)malloc(len)))
		return -1;
	for (i = 0; len > i; i++)
		if (1 != sscanf(&in[i * 2], "%2hhx", *out + i) ||
		    (!ishexnumber(in[(i * 2) + 1]))) {
			free(*out);
			return -1;
		}
	return 0;
}

static void
usage(void)
{
	fprintf(stderr,
	    "usage: ocra_tool init -f credential_file -k key -s suite_string\n"
	    "                     [-c counter] [-p pin | -P pin_hash]\n"
	    "                     [-q kill_pin | -Q kill_pin_hash]\n"
	    "                     [-w counter_window] [-t timestamp_offset]\n"
	    "       ocra_tool info -f credential_file\n"
	    "       ocra_tool sync -f credential_file -c challenge\n"
	    "                      -r response -v second_response\n");
	exit(-1);
}

static void
cmd_info(int argc, char **argv)
{
	int ch, ret;
	uint32_t i;
	char *fname = NULL;
	ocra_suite ocra;

	DB *db;
	DBT K, V;

	while (-1 != (ch = getopt(argc, argv, "f:"))) {
		switch (ch) {
		case 'f':
			if (NULL != fname)
				usage();
			fname = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	if ((0 != argc) ||
	    (NULL == fname))
		usage();

	memset(&K, 0, sizeof(K));
	memset(&V, 0, sizeof(V));

	if (NULL ==
	    (db = dbopen(fname, O_EXLOCK | O_RDONLY, 0600, DB_BTREE, NULL)))
		err(EX_OSERR, "dbopen() failed");

	KEY(K, "suite");
	if (0 != (ret = db->get(db, &K, &V, 0)))
		errx(EX_OSERR, "db->get() failed: %s",
		    (1 == ret) ? "key not in db" : strerror(errno));
	printf("suite:\t\t%s\n", (char *)(V.data));

	if (RFC6287_SUCCESS != (ret = rfc6287_parse_suite(&ocra, V.data)))
		errx(EX_SOFTWARE, "rfc6287_parse_suite() failed: %s",
		    rfc6287_err(ret));

	KEY(K, "key");
	if (0 != (ret = db->get(db, &K, &V, 0)))
		errx(EX_OSERR, "db->get() failed: %s",
		    (1 == ret) ? "key not in db" : strerror(errno));
	if (mdlen(ocra.hotp_alg) != V.size)
		errx(EX_SOFTWARE, "key size does not match suite!");

	printf("key:\t\t0x");
	for (i = 0; V.size > i; i++)
		printf("%02x", ((uint8_t *)(V.data))[i]);
	printf("\n");

	if (ocra.flags & FL_C) {
		uint64_t C;
		int CW;

		KEY(K, "C");
		if (0 != (ret = db->get(db, &K, &V, 0)))
			errx(EX_OSERR, "db->get() failed: %s",
			    (1 == ret) ? "key not in db" : strerror(errno));
		memcpy(&C, V.data, sizeof(C));
		printf("counter:\t0x%.16" PRIx64 "\n", C);

		KEY(K, "counter_window");
		if (0 != (ret = db->get(db, &K, &V, 0)))
			errx(EX_OSERR, "db->get() failed: %s",
			    (1 == ret) ? "key not in db" : strerror(errno));
		memcpy(&CW, V.data, sizeof(CW));
		printf("counter_window: %d\n", CW);
	}
	if (ocra.flags & FL_P) {
		int kill_pin_used = 0;

		KEY(K, "P");
		if (0 != (ret = db->get(db, &K, &V, 0)))
			errx(EX_OSERR, "db->get() failed: %s",
			    (1 == ret) ? "key not in db" : strerror(errno));

		if (mdlen(ocra.P_alg) != V.size)
			errx(EX_SOFTWARE, "pin hash size does not match suite!");
		printf("pin_hash:\t0x");
		for (i = 0; V.size > i; i++)
			printf("%02x", ((uint8_t *)(V.data))[i]);
		printf("\n");

		KEY(K, "kill_pin");
		ret = db->get(db, &K, &V, 0);
		if (ret != 0 && ret != 1)
			errx(EX_OSERR, "db->get() failed: %s", strerror(errno));
		if (ret == 1 || V.size == 0)
			printf("kill_pin_hash:\tNOT SET\n");
		else {
			if (mdlen(ocra.P_alg) != V.size) {
				errx(EX_SOFTWARE,
				    "kill pin hash size does not match suite!");
			}
			printf("kill_pin_hash:\t0x");
			for (i = 0; V.size > i; i++)
				printf("%02x", ((uint8_t *)(V.data))[i]);

			printf("\n");
		}

		KEY(K, "kill_pin_used");
		if (0 == (ret = db->get(db, &K, &V, 0))) {
			memcpy(&kill_pin_used, V.data, sizeof(kill_pin_used));
		} else if (ret != 1)
			errx(EX_OSERR, "db->get() failed: %s", strerror(errno));

		printf("kill pin used:\t%s\n",
		    (ret == 1 || !kill_pin_used) ? "false" : "true");
	}
	if (ocra.flags & FL_T) {
		int TO;

		KEY(K, "timestamp_offset");
		if (0 != (ret = db->get(db, &K, &V, 0)))
			errx(EX_OSERR, "db->get() failed: %s",
			    (1 == ret) ? "key not in db" : strerror(errno));
		memcpy(&TO, V.data, sizeof(TO));
		printf("timestamp_offset: %d\n", TO);
	}
	if (0 != (db->close(db)))
		errx(EX_OSERR, "db->close() failed: %s", strerror(errno));
}

static void
test_input(const ocra_suite * ocra, const char *suite_string,
    const uint8_t *key, size_t key_l, uint64_t C, const uint8_t *P,
    size_t P_l, int counter_window, int timestamp_offset)
{
	int r;
	uint64_t T;
	uint64_t next_counter;
	char *questions;
	char *response;

	if (RFC6287_SUCCESS != (r = rfc6287_challenge(ocra, &questions)))
		errx(EX_SOFTWARE, "rfc6287_challenge() failed: %s",
		    rfc6287_err(r));

	if (RFC6287_SUCCESS != (r = rfc6287_timestamp(ocra, &T)))
		errx(EX_SOFTWARE, "rfc6287_timestamp() failedi: %s",
		    rfc6287_err(r));

	if (RFC6287_SUCCESS != (r = rfc6287_ocra(ocra, suite_string,
	    key, key_l, C, questions, P, P_l, NULL, 0, T, &response)))
		errx(EX_SOFTWARE, "rfc6287_ocra() failed: %s", rfc6287_err(r));

	if (RFC6287_SUCCESS != (r = rfc6287_verify(ocra, suite_string,
	    key, key_l, C, questions, P, P_l, NULL, 0, T, response,
	    counter_window, &next_counter, timestamp_offset)))
		errx(EX_SOFTWARE, "rfc6287_verify() failed: %s",
		    rfc6287_err(r));

	free(response);
	free(questions);
}

static void
write_db(const char *fname, const char *suite_string,
    const uint8_t *key, size_t key_l, uint64_t C, const uint8_t *P,
    size_t P_l, const uint8_t *KP, size_t KP_l, int counter_window,
    int timestamp_offset)
{
	DB *db;
	DBT K, V;

	memset(&K, 0, sizeof(K));
	memset(&V, 0, sizeof(V));

	if (NULL == (db = dbopen(fname, O_CREAT | O_EXLOCK | O_RDWR | O_TRUNC,
	    0600, DB_BTREE, NULL)))
		err(EX_OSERR, "dbopen() failed");

	KEY(K, "suite");
	VALUE(V, suite_string, strlen(suite_string) + 1);
	if (0 != (db->put(db, &K, &V, R_NOOVERWRITE)))
		err(EX_OSERR, "db->put() failed");

	KEY(K, "key");
	VALUE(V, key, key_l);
	if (0 != (db->put(db, &K, &V, R_NOOVERWRITE)))
		err(EX_OSERR, "db->put() failed");

	KEY(K, "C");
	VALUE(V, &C, sizeof(C));
	if (0 != (db->put(db, &K, &V, R_NOOVERWRITE)))
		err(EX_OSERR, "db->put() failed");

	KEY(K, "P");
	VALUE(V, P, P_l);
	if (0 != (db->put(db, &K, &V, R_NOOVERWRITE)))
		err(EX_OSERR, "db->put() failed");

	KEY(K, "kill_pin");
	VALUE(V, KP, KP_l);
	if (0 != (db->put(db, &K, &V, R_NOOVERWRITE)))
		err(EX_OSERR, "db->put() failed");

	KEY(K, "counter_window");
	VALUE(V, &counter_window, sizeof(counter_window));
	if (0 != (db->put(db, &K, &V, R_NOOVERWRITE)))
		err(EX_OSERR, "db->put() failed");

	KEY(K, "timestamp_offset");
	VALUE(V, &timestamp_offset, sizeof(timestamp_offset));
	if (0 != (db->put(db, &K, &V, R_NOOVERWRITE)))
		err(EX_OSERR, "db->put() failed");

	if (0 != (db->sync(db, 0)))
		err(EX_OSERR, "db->sync() failed");

	if (0 != (db->close(db)))
		err(EX_OSERR, "db->close() failed");
}

static void
cmd_init(int argc, char **argv)
{
	int r;
	int ch;
	char *fname = NULL;
	char *suite_string = NULL;
	char *key_string = NULL;
	char *pin_string = NULL;
	char *pin_hash_string = NULL;
	char *kill_pin_string = NULL;
	char *kill_pin_hash_string = NULL;
	char *counter_string = NULL;
	char *counter_window_string = NULL;
	char *timestamp_offset_string = NULL;

	ocra_suite ocra;
	uint64_t C = 0;
	int timestamp_offset = 0;
	int counter_window = 0;

	uint8_t *KP = NULL;
	size_t KP_l = 0;
	uint8_t *P = NULL;
	size_t P_l = 0;
	uint8_t *key = NULL;
	size_t key_l = 0;

	while (-1 != (ch = getopt(argc, argv, "f:s:k:p:P:c:w:t:q:Q:"))) {
		switch (ch) {
		case 'f':
			if (NULL != fname)
				usage();
			fname = optarg;
			break;
		case 's':
			if (NULL != suite_string)
				usage();
			suite_string = optarg;
			break;
		case 'k':
			if (NULL != key_string)
				usage();
			key_string = optarg;
			break;
		case 'c':
			if (NULL != counter_string)
				usage();
			counter_string = optarg;
			break;
		case 'p':
			if (NULL != pin_string)
				usage();
			pin_string = optarg;
			break;
		case 'P':
			if (NULL != pin_hash_string)
				usage();
			pin_hash_string = optarg;
			break;
		case 'q':
			if (NULL != kill_pin_string)
				usage();
			kill_pin_string = optarg;
			break;
		case 'Q':
			if (NULL != kill_pin_hash_string)
				usage();
			kill_pin_hash_string = optarg;
			break;
		case 'w':
			if (NULL != counter_window_string)
				usage();
			counter_window_string = optarg;
			break;
		case 't':
			if (NULL != timestamp_offset_string)
				usage();
			timestamp_offset_string = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	if ((0 != argc) ||
	    (NULL == fname) ||
	    (NULL == suite_string) ||
	    (NULL == key_string))
		usage();

	if (RFC6287_SUCCESS != (r = rfc6287_parse_suite(&ocra, suite_string)))
		err(EX_CONFIG, "rfc6287_parse_suite() failed: %s",
		    rfc6287_err(r));

	if (ocra.flags & FL_C) {
		if (NULL == counter_string)
			errx(EX_CONFIG, "suite requires counter parameter "
			    "(-c <counter> missing)");
		if (-1 == parse_counter(counter_string, &C))
			errx(EX_CONFIG, "invalid counter value");
		if (NULL != counter_window_string)
			if (-1 ==
			    (counter_window = parse_num(counter_window_string)))
				errx(EX_CONFIG, "invalid counter window value");
	} else {
		if (NULL != counter_string)
			errx(EX_CONFIG, "suite does not require counter "
			    "parameter (-c <counter> must not be set)");
		if (NULL != counter_window_string)
			errx(EX_CONFIG, "suite does not require counter "
			    "parameter  (-w <counter_window> must not be set)");
	}

	if (ocra.flags & FL_S)
		errx(EX_CONFIG, "suite requires session parameter (S) which"
		    " is not supported by pam_ocra");

	if (ocra.flags & FL_T) {
		if (-1 ==
		    (timestamp_offset = parse_num(timestamp_offset_string)))
			errx(EX_CONFIG, "invalid timestamp offset value");
	} else if (NULL != timestamp_offset_string)
		errx(EX_CONFIG, "suite does nor require timestamp parameter "
		    " (-t <timestamp_offset> must not be set)");

	if (0 == ocra.hotp_trunc)
		errx(EX_CONFIG, "suite specifies no (0) truncation in "
		    "CryptoFunction. This is not supported by pam_ocra");

	if (ocra.flags & FL_P) {
		if (NULL != pin_string && NULL != pin_hash_string)
			errx(EX_CONFIG, "exactly one of -p <pin> and -P "
			    "<pin_hash> must be set");
		if (NULL != pin_string)
			pin_hash(&ocra, pin_string, &P, &P_l);
		else if (NULL != pin_hash_string) {
			P_l = mdlen(ocra.P_alg);
			if (0 != from_hex(pin_hash_string, &P, P_l))
				errx(EX_CONFIG, "invalid pin_hash");
		} else
			errx(EX_CONFIG, "suite requires pin parameter "
			    "(-p <pin> or -P <pin_hash> missing)");
		if (NULL != kill_pin_string && NULL != kill_pin_hash_string)
			errx(EX_CONFIG, "only one of -q <kill_pin> or -Q "
			    "<kill_pin_hash> must be set");
		if (NULL != kill_pin_string)
			pin_hash(&ocra, kill_pin_string, &KP, &KP_l);
		if (NULL != kill_pin_hash_string) {
			KP_l = mdlen(ocra.P_alg);
			if (0 != from_hex(kill_pin_hash_string, &KP, KP_l))
				errx(EX_CONFIG, "invalid kill_pin_hash");
		}
	} else if (NULL != pin_string || NULL != pin_hash_string ||
	    NULL != kill_pin_string || NULL != kill_pin_hash_string)
		errx(EX_CONFIG, "suite does not require pin parameter"
		    " (-p <pin> and -P <pin_hash> must not be set)");

	key_l = mdlen(ocra.hotp_alg);
	if (0 != from_hex(key_string, &key, key_l))
		errx(EX_CONFIG, "invalid key");

	test_input(&ocra, suite_string, key, key_l, C, P, P_l,
	    counter_window, timestamp_offset);

	write_db(fname, suite_string, key, key_l, C, P, P_l, KP, KP_l,
	    counter_window, timestamp_offset);

}

static void
cmd_sync_counter(int argc, char **argv)
{
	int r;
	int ch;
	char *fname = NULL;
	char *challenge = NULL;
	char *response1 = NULL;
	char *response2 = NULL;
	char *suite_string = NULL;

	ocra_suite ocra;
	int timestamp_offset = 0;
	uint8_t *P = NULL;
	uint8_t *key = NULL;
	size_t P_l = 0;
	size_t key_l = 0;
	uint64_t C = 0;
	uint64_t T = 0;

	DB *db = NULL;
	DBT K, V;

	memset(&K, 0, sizeof(K));
	memset(&V, 0, sizeof(V));

	while (-1 != (ch = getopt(argc, argv, "f:c:r:v:"))) {
		switch (ch) {
		case 'f':
			if (NULL != fname)
				usage();
			fname = optarg;
			break;
		case 'c':
			if (NULL != challenge)
				usage();
			challenge = optarg;
			break;
		case 'r':
			if (NULL != response1)
				usage();
			response1 = optarg;
			break;
		case 'v':
			if (NULL != response2)
				usage();
			response2 = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	if ((0 != argc) || (NULL == fname) || (NULL == challenge) ||
	    (NULL == response1) || (NULL == response2))
		usage();

	if (NULL == (db = dbopen(fname, O_EXLOCK | O_RDWR, 0, DB_BTREE, NULL)))
		err(EX_OSERR, "dbopen() failed");

	KEY(K, "suite");
	if (0 != (r = db->get(db, &K, &V, 0)))
		errx(EX_OSERR, "db->get() failed: %s",
		    (1 == r) ? "key not in db" : strerror(errno));
	if (NULL == (suite_string = (char *)malloc(V.size)))
		errx(EX_OSERR, "malloc() failed: %s", strerror(errno));
	memcpy(suite_string, V.data, V.size);

	if (RFC6287_SUCCESS != (r = rfc6287_parse_suite(&ocra, V.data)))
		errx(EX_SOFTWARE, "rfc6287_parse_suite() failed: %s",
		    rfc6287_err(r));

	if (0 == (ocra.flags & FL_C))
		errx(EX_CONFIG, "suite does not require counter, can not sync");

	if (ocra.Q_l != (int)strlen(challenge))
		errx(EX_CONFIG, "challenge length does not match suite (%d)",
		    ocra.Q_l);

	if ((ocra.hotp_trunc != (int)strlen(response1)) ||
	    (ocra.hotp_trunc != (int)strlen(response2)))
		errx(EX_CONFIG, "response length does not match suite (%d)",
		    ocra.hotp_trunc);

	KEY(K, "key");
	if (0 != (r = db->get(db, &K, &V, 0)))
		errx(EX_OSERR, "db->get() failed: %s",
		    (1 == r) ? "key not in db" : strerror(errno));
	if (NULL == (key = (uint8_t *)malloc(V.size)))
		errx(EX_OSERR, "malloc() failed: %s", strerror(errno));
	memcpy(key, V.data, V.size);
	key_l = V.size;

	if (ocra.flags & FL_P) {
		KEY(K, "P");
		if (0 != (r = db->get(db, &K, &V, 0)))
			errx(EX_OSERR, "db->get() failed: %s",
			    (1 == r) ? "key not in db" : strerror(errno));
		if (NULL == (P = (uint8_t *)malloc(V.size)))
			errx(EX_OSERR, "malloc() failed: %s", strerror(errno));
		memcpy(P, V.data, V.size);
		P_l = V.size;
	}
	if (ocra.flags & FL_T) {
		KEY(K, "timestamp_offset");
		if (0 != (r = db->get(db, &K, &V, 0)))
			errx(EX_OSERR, "db->get() failed: %s",
			    (1 == r) ? "key not in db" : strerror(errno));
		memcpy(&timestamp_offset, V.data, sizeof(timestamp_offset));

		if (0 != (r = rfc6287_timestamp(&ocra, &T)))
			errx(EX_SOFTWARE, "rfc6287_timestamp() failed: %s",
			    rfc6287_err(r));
	}
	while (C < UINT64_MAX) {
		r = rfc6287_verify(&ocra, suite_string, key, key_l, C,
		    challenge, P, P_l, NULL, 0, T, response1, INT32_MAX,
		    &C, timestamp_offset);
		if (r != RFC6287_SUCCESS)
			continue;

		printf("candidate:\t0x%.16" PRIx64 "\n", C);
		r = rfc6287_verify(&ocra, suite_string, key, key_l, C,
		    challenge, P, P_l, NULL, 0, T, response2, 1, &C,
		    timestamp_offset);

		if (r == RFC6287_SUCCESS) {
			printf("match:\t\t0x%.16" PRIx64 "\n", C);
			KEY(K, "C");
			VALUE(V, &C, sizeof(C));
			if (0 != (db->put(db, &K, &V, 0)))
				err(EX_OSERR, "db->put() failed");
			if (0 != (db->sync(db, 0)))
				err(EX_OSERR, "db->sync() failed");
			break;
		}
	}

	if (0 != (db->close(db)))
		err(EX_OSERR, "db->close() failed");
	free(suite_string);
	free(key);
	if (P != NULL)
		free(P);
}

int
main(int argc, char **argv)
{
	if (2 > argc)
		usage();
	if (0 == strcmp(argv[1], "init"))
		cmd_init(argc - 1, argv + 1);
	else if (0 == strcmp(argv[1], "info"))
		cmd_info(argc - 1, argv + 1);
	else if (0 == strcmp(argv[1], "sync"))
		cmd_sync_counter(argc - 1, argv + 1);
	else
		usage();
	return 0;
}
